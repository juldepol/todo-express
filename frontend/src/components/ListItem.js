import React, { Component } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes, faCheck } from '@fortawesome/free-solid-svg-icons';

class ListItem extends Component {
    /**
     * Renders complete button if task is not completed yet
     */
    completeButton(){
        if (!this.props.data.done){
            return (
                <button className="btn btn-lg completeBtn"
                        onClick={this.props.completeTask}>
                    <FontAwesomeIcon icon={faCheck} />
                </button>
            );
        } 
    }
    
    render() {
        return (
            <div className="todoListItem">
                <div className="col-md-8 offset-md-2">
                    <div className="row">
                        <div className="col-md-10 col-sm-8">
                            <h3 className={(this.props.data.done) ? "taskText taskDone" : "taskText"}>
                                {this.props.data.task}
                            </h3>
                        </div>
                        <div className="col-md-1 col-sm-2">
                            {this.completeButton()}
                        </div>
                        <div className="col-md-1 col-sm-2">
                            <button className="btn btn-lg float-right deleteBtn"
                                    onClick={this.props.deleteTask}>
                                <FontAwesomeIcon icon={faTimes} />
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}
export default ListItem