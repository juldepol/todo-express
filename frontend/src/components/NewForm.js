import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus  } from '@fortawesome/free-solid-svg-icons';

class NewForm extends Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }
    
    /**
     * Handles form submit
     * @param {Event} e 
     */
    onSubmit(e) {
        e.preventDefault();
        var input = ReactDOM.findDOMNode(this.refs.newTask);
        var newTask = input.value;
        if(newTask) {
            this.props.addTask({task: 
                newTask,
                done: false
            });
            input.value = '';
        }
    }

    render() {
        return (
        <div className="addNewtForm">
            <form onSubmit={this.onSubmit}>
                <input type="text" placeholder=" Task" ref="newTask" />
                <button className="btn btn-lg" type="submit">
                    <FontAwesomeIcon icon={faPlus} />
                </button>
            </form>
        </div>
        )
    }
}
export default NewForm