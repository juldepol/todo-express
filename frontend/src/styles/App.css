.App {
  background-color: #fffaf0;
  font-family: Console, monospace;
}

.header{
  margin-top: 20px;
  margin-bottom: 10px;
  font-size: 30px;
}

.addNewtForm{
  text-align: center;
  font-size: 20px;
}

.taskText{
  margin-top: 10px;
  font-size: 20px;
}

.taskDone{
  text-decoration: line-through;
}

.btn:focus,.btn:active{
  outline: none !important;
  box-shadow: none !important;
}

.btn:hover{
  color: gray !important;
}

input[type="text"]{
  background-color : #fffaf0; 
  outline: none !important;
  box-shadow: none !important;
  border: black 2px solid;
  border-radius: 5px;
}