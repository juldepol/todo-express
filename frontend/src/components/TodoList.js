import React, { Component } from 'react';
import ListItem from './ListItem';

class TodoList extends Component {
    render() {
        return (
            <div className="todoListMain">
                <div className="col-md-12">
                {(this.props.list.length === 0) ? 
                    <h4 className="text-center">List is empty</h4> : ''
                }
                {this.props.list.map((value, index) => {
                    return <ListItem key={index}
                        data={value}
                        completeTask={() => this.props.completeTask(index)}
                        deleteTask={() => this.props.deleteTask(index)}/>
                })}  
                </div> 
            </div>
        )
    }
}
export default TodoList