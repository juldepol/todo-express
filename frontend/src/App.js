import React, {Component} from 'react';
import TodoList from './components/TodoList';
import NewForm from './components/NewForm';
import axios from "axios"
import './styles/App.css';

/**
 * General component
 */
class App extends Component {
  constructor(props){
    super(props);
    this.state = {
      list: []
    };
    this.handleDelete = this.handleDelete.bind(this);
    this.handleComplete = this.handleComplete.bind(this);
    this.handleAdd = this.handleAdd.bind(this);
  }

  /**
  * Fetches list of todos from server
  */
  getList(){
    axios.get("/task").then(results => {
      return results;
    }).then(data => {
        let tasks = data.data.map((task) => {
          return {
            id: task.id,
            task: task.task,
            done: task.done
          }
        });
        this.setState({list: tasks});
    });
  }

  componentDidMount(){
    this.getList();
  }

  /**
   * Submits new task to server and updates list
   * @param {Object} item 
   */
  handleAdd(item) {
    axios.post("/task", {
      task: item.task
    }).catch(function (error) {
      console.log(error);
    });
    this.getList();
  }

  /**
   * Sends delete request to server, updates state 
   * @param {int} id 
   */
  handleDelete(id) {
    var self = this;
    let tasks = [...this.state.list];
    tasks.forEach((element, index) => {
      if (index === id){
        axios.delete("/task?id="+element.id)
        .then(function (response) {
          if (response.status === 200){
            self.setState({list: tasks.filter((element, index) => {
              return index !== id;
            })});
          }
        }).catch(function (error) {
          console.log(error);
        });
      }
    });
  }

  /**
   * Sends request to server to make selected task completed, updates state 
   * @param {int} id 
   */
  handleComplete(id) {
    var self = this;
    let tasks = [...this.state.list];
    tasks[id].done = true;
    axios.put("/task?id="+tasks[id].id)
    .then(function (response) {
      if (response.status === 200){
        self.setState({list: tasks});
      }
    }).catch(function (error) {
      console.log(error);
    });
  }

  render() {
    return (<div className="App">
      <div className="header">
        <div className="col-md-12 col-sm-12">
          <h3 className="text-center">ToDo List</h3>
        </div>
      </div>
      <NewForm addTask={this.handleAdd}/>
      <TodoList list={this.state.list}
      completeTask={id => this.handleComplete(id)}
      deleteTask={id => this.handleDelete(id)}/>
    </div>)
  };
}

export default App;
