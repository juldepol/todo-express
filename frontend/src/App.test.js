import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { render } from 'react-testing-library';
import 'react-testing-library/cleanup-after-each';
import 'jest-dom/extend-expect'
import { shallow, configure } from 'enzyme'
import toJson from 'enzyme-to-json'
import ListItem from './components/ListItem'
import NewForm from './components/NewForm'
import TodoList from './components/TodoList'
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders page header', () => {
    const { getByText } = render(<App />);
    expect(getByText('ToDo List')).toBeInTheDocument();
  });
});

describe('NewForm', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<NewForm />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('input is cleared after submit', () => {
    const fakeEvent = { 
      preventDefault: () => console.log('preventDefault'),
      
    };
    const NewFormComponent = shallow(<NewForm />);
    NewFormComponent.find('input').simulate('submit', {value: "test"});
    console.log(Object.keys(NewFormComponent.find('input')));
    expect(NewFormComponent.find('input')).toEqual({});
  });
});

describe('TodoList', () => {
  it('renders without crashing given the required props', () => {
    const props = {
      list: [
        {
          id: 0, 
          task: "test 1", 
          done: 0
        }, {
          id: 1, 
          task: "test 2", 
          done: 1
        }]
    }
    const wrapper = shallow(<TodoList {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

  it('understands if the list is empty', () => {
    const props = {
      list: []
    }
    const wrapper = shallow(<TodoList {...props} />)
    const header = wrapper.find('h4.text-center')
    expect(header.text().includes('List is empty')).toBe(true);
  });
});

describe('ListItem', () => {
  it('renders without crashing given the required props', () => {
    const props = {
      key: 0,
      data: {
        id: 0,
        task: "test task",
        done: 1
      }
    }
    const wrapper = shallow(<ListItem {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });

  it('renders without complete button and crossed text when task is completed', () => {
    const props = {
      key: 0,
      data: {
        id: 0,
        task: "test task",
        done: 1
      }
    }
    const wrapper = shallow(<ListItem {...props} />)
    expect(wrapper.find('.taskDone')).toBeDefined();
    console.log(wrapper.find('.taskDone'));
    expect(wrapper.find('.completeBtn')).toEqual({});
  });

  it('renders with complete button when task is not completed', () => {
    const props = {
      key: 0,
      data: {
        id: 0,
        task: "test task",
        done: 0
      }
    }
    const wrapper = shallow(<ListItem {...props} />)
    expect(wrapper.find('.completeBtn')).toBeDefined();
  });

  it('renders with delete button', () => {
    const props = {
      key: 0,
      data: {
        id: 0,
        task: "test task",
        done: 0
      }
    }
    const wrapper = shallow(<ListItem {...props} />)
    expect(wrapper.find('.deleteBtn')).toBeDefined();
  });
});

