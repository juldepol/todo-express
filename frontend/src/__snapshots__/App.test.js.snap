// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders without crashing given the required props 1`] = `
<div
  className="todoListItem"
>
  <div
    className="col-md-8 offset-md-2"
  >
    <div
      className="row"
    >
      <div
        className="col-md-10 col-sm-8"
      >
        <h3
          className="taskText taskDone"
        >
          test task
        </h3>
      </div>
      <div
        className="col-md-1 col-sm-2"
      />
      <div
        className="col-md-1 col-sm-2"
      >
        <button
          className="btn btn-lg float-right deleteBtn"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  352,
                  512,
                  Array [],
                  "f00d",
                  "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
                ],
                "iconName": "times",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoList renders without crashing given the required props 1`] = `
<div
  className="todoListMain"
>
  <div
    className="col-md-12"
  >
    <ListItem
      completeTask={[Function]}
      data={
        Object {
          "done": 0,
          "id": 0,
          "task": "test 1",
        }
      }
      deleteTask={[Function]}
      key="0"
    />
    <ListItem
      completeTask={[Function]}
      data={
        Object {
          "done": 1,
          "id": 1,
          "task": "test 2",
        }
      }
      deleteTask={[Function]}
      key="1"
    />
  </div>
</div>
`;
